src/agent/Core/ApplicationPool/Session.h

```diff
@@ -186,19 +186,21 @@ class Session: public AbstractSession {
 	}
 
 
-	virtual void initiate(bool blocking = true) {
+	virtual void initiate(bool blocking = false) {
 		assert(!closed);
 		ScopeGuard g(boost::bind(&Session::callOnInitiateFailure, this));
 		Connection connection = socket->checkoutConnection();
 		connection.fail = true;
-		if (connection.blocking && !blocking) {
-			FdGuard g2(connection.fd, NULL, 0);
-			setNonBlocking(connection.fd);
-			g2.clear();
-			connection.blocking = false;
-		}
+		FdGuard g2(connection.fd, NULL, 0);
+		setNonBlocking(connection.fd);
+		g2.clear();
 		g.clear();
 		this->connection = connection;
+		if (blocking) {
+			// for testing purposes
+			if (!connection.ready) { connection.wait(); }
+			setBlocking(connection.fd);
```

Comment:
If wait or setBlocking fails here then the connection won't be cleaned up since g2 is already cleared.

But now that I think about it: since we *only* care about blocking I/O in the tests, I think it's better to move the wait to the tests. Just don't have blocking I/O support in ApplicationPool at all, and have only the tests deal with that.